File: exercices/exchange.c
#include <stdio.h>
#include <mpi.h>

int main(int argc, char* argv[])
{
    int rank, size;
    float A[2], B[2];
    MPI_Status status;

    MPI_Init( &argc, &argv );
    MPI_Comm_rank( MPI_COMM_WORLD, &rank );
    MPI_Comm_size( MPI_COMM_WORLD, &size );
    if (size < 2) { MPI_Finalize(); return -1; }

    if (rank == 0) {
        A[0] = 1.0f;
        A[1] = 2.0f;

        // Send and Receive
        MPI_...( ... );
        MPI_...( ... );
        printf("Process %d sent     A=[%f, %f]\n", rank, A[0], A[1]);
        printf("Process %d received B=[%f, %f]\n", rank, B[0], B[1]);
    }

    if (rank == 1) {
        B[0] = 3.0f;
        B[1] = 4.0f;

        // Send and Receive
        MPI_...( ... );
        MPI_...( ... );
        printf("Process %d sent     B=[%f, %f]\n", rank, B[0], B[1]);
        printf("Process %d received A=[%f, %f]\n", rank, A[0], A[1]);
    }

    MPI_Finalize();
    return 0;
}

File: exercices/exchange.f90
PROGRAM exchange
use mpi
IMPLICIT NONE

INTEGER :: rank
REAL :: A(2), B(2)
INTEGER :: error, status(MPI_STATUS_SIZE)

    CALL MPI_Init( error )
    CALL MPI_Comm_rank( MPI_COMM_WORLD, rank, error )

    IF (rank == 0) THEN
        A=(/1.0,2.0/)

        CALL ...
        CALL ...
        WRITE(*,*) 'Process', rank, 'sent     A=', A
        WRITE(*,*) 'Process', rank, 'received B=', B

    ELSE IF (rank == 1) THEN
        B=(/3.0,4.0/)

        CALL ...
        CALL ...
        WRITE(*,*) 'Process', rank, 'sent     B=', B
        WRITE(*,*) 'Process', rank, 'received A=', A
    END IF

    CALL MPI_Finalize( error )

END PROGRAM exchange

